hpct-ldap-client/src/managers/openldap.py

```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+#
+# managers/ldap.py
+"""Manager for client operator."""
+import logging
+
+logger = logging.getLogger(__name__)
+import subprocess
+
+from charms.operator_libs_linux.v0 import apt
+from charms.operator_libs_linux.v1 import systemd
+
+from utils.filedata import FileData
+
+
+class OpenldapClientManager:
+    """Manager to provide LDAP client charm all functionality needed."""
+
+    packages = ["ldap-utils", "sssd-ldap"]
+    systemd_services = ["sssd"]
+
+    def __init__(self):
+        pass
+
+    def disable(self) -> None:
+        """Disable services."""
+        for name in self.systemd_services:
+            systemd.service_pause(name)
+
+    def enable(self) -> None:
+        """Enable services."""
+        for name in self.systemd_services:
+            systemd.service_resume(name)
+
+    def install(self) -> None:
+        """Install using charmlib apt."""
+        if self.packages:
+            try:
+                apt.update()
+                for name in self.packages:
+                    apt.add_package(name)
+            except:
+                raise Exception(f"failed to install package ({name})")
+
+    def is_enabled(self) -> bool:
+        """Check enabled status of services."""
+        if self.systemd_services:
+            for name in self.systemd_services:
+                if not systemd._systemctl("is-enabled", name, quiet=True):
+                    return False
+
+        return True
+
+    def is_installed(self) -> bool:
+        """Check packages are installed."""
+        if self.packages:
+            for name in self.packages:
+                if not apt.DebianPackage.from_installed_package(name).present:
+                    return False
```

Comment:
This is what I had to do to avoid a possible exception.

```
    def is_installed(self):
        if self.packages:
            for name in self.packages:
                try:
                    if not apt.DebianPackage.from_installed_package(name).present:
                        return False
                except:
                    return False
        return True
```

If you call the `is_installed()` before the `install`, you'll likely encounter the problem with your code, too. Please verify.

Reply:
I did encounter a problem and that's why I put the `is_installed` check after the `install` thinking I misunderstood your previous comment and that's why I was seeing an error. I see now that I need to make this update and put the `is_installed` check before the `install` will do now.

Reply:
Updated and works fine now as suggested